### PR TITLE
Enforce limit of NSIG signals.

### PR DESCRIPTION
**Pull request by stoeckmann**
It doesn't make sense to allow a signal number higher than NSIG.
At worst, libevent could enter an endless loop if sig is INT_MAX.

The NSIG check already exists in signal.c for internally used
functions.

As this is a programming error of libevent consumers, this is a
purely defensive mechanism.

Example to trigger this issue:

--poc.c--

static void callback(int a, short b, void *c) { }

int main(void)
{
  struct event_base *b = event_init();
  struct event *ev = evsignal_new(b, INT_MAX, callback, (void *)b);
  event_add(ev, NULL);
  return 0;
}
--poc.c--

Signed-off-by: Tobias Stoeckmann <tobias@stoeckmann.org>

### CODE REVIEW

**Comment by azat** [comment]
Nice catch!

But assert for this case is not a good thing IMO, since this will bail only in debug mode, but it **should bail always**

And I don't think that it is worth expanding `event_signal_map.nentries` to 8 bytes, since it will not be any sane/optimal to have that much of fds, and also I'm not sure about that hashtable.

So what I'm suggesting is decline fds greater then `INT_MAX` but be aware of win32, which has pointer-like fds instead of lowest possible number in POSIX compatible os. So for win32 we should not have this restrictions for regular event fds, but in all other cases we should decline then with an error (and also it will be great to fix endless loop in evmap_make_space() due to this overflow)

**Comment by stoeckmann** [comment]
Oh, thanks for clarification about the asserts. I've got too used to tor's asserts which are basically always around. :)

I will refactor the patch based on your suggestions!

**Comment by azat** [comment]
> Oh, thanks for clarification about the asserts. I've got too used to tor's asserts which are basically always around. :)

Yeah, asserts for things that "should never happen", like passing `NULL` pointers where they are not expected, but sometimes this checks is too costly and here comes `NDEBUG` macros, which turns off asserts

*P.S. [here](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=711515) you can find some discussion from debian bug tracker about should packages be built with `NDEBUG` or not.*

**Comment by azat** [comment]
Also this can be covered within regression tests

**Comment by stoeckmann** [comment]
I have adjusted the merge request accordingly:
- I still enforce an upper limit of NSIG to stay in sync with internal code
- The enforcement is not an assert any longer but a hard failure
- evmap_make_space cannot loop endlessly on exceptionally large fds anymore
- While at it I fixed an out of boundary access on negative signals (during deletion, if at all exploitable)
- Extended tests to cover these cases, except the sig deletion part

Windows is not affected and doesn't take any special treatment, because Windows uses the hash map. So, from my understanding, all points you mentioned should have been fixed by now.

Looking forward to get feedback, especially if these test looks good to you.

File: test/regress.c [hunk 1]
```diff
@@ -3303,6 +3303,46 @@ tabf_cb(evutil_socket_t fd, short what, void *arg)
...
+	ev1 = evsignal_new(base, -1, dummy_read_cb, (void *)base);
+	ev2 = evsignal_new(base, NSIG, dummy_read_cb, (void *)base);
+	tt_assert(ev1);
+	tt_assert(ev2);
+	e1 = event_add(ev1, NULL);
+	e2 = event_add(ev2, NULL);
+	tt_int_op(e1, !=, 0);
+	tt_int_op(e2, !=, 0);
+#ifndef _WIN32
+	ev3 = event_new(base, INT_MAX, EV_READ, dummy_read_cb, (void *)base);
```
Comment (azat) [inline test/regress.c]: Missing `limits.h` for `INT_MAX` (https://travis-ci.org/libevent/libevent/jobs/525026817#L1187)

Comment (stoeckmann) [reply test/regress.c]: Fixed. Missed that because it still worked on my Linux box.

**Comment by azat** [comment]
Applied, thanks!

**Pull request merged**